packages/SwingSet/src/kernel/state/vatKeeper.js

```diff
@@ -567,6 +567,27 @@ export function makeVatKeeper(
     return true;
   }
 
+  function removeSnapshotAndTranscript() {
+    const skey = `local.${vatID}.lastSnapshot`;
+    const epkey = `${vatID}.t.endPosition`;
+    if (snapStore) {
+      const notation = kvStore.get(skey);
+      if (notation) {
+        const { snapshotID } = JSON.parse(notation);
+        if (removeFromSnapshot(snapshotID) === 0) {
+          // TODO: if we roll back (because the upgrade failed), we must
+          // not really delete the snapshot
+          snapStore.prepareToDelete(snapshotID);
+        }
+        kvStore.delete(skey);
+      }
+    }
+    // TODO: same rollback concern
+    // TODO: streamStore.deleteStream(transcriptStream);
+    const newStart = streamStore.STREAM_START;
```

Comment:
Do we really want to reset the stream pointer to the start position or merely remove the prior transcript entries?  Seems like for debugging purposes it would be nicer to maintain the continuity of transcript position across the upgrade, so that, for example, if one was looking at a historical record of a transcript there would be no ambiguity about which incarnation of the vat code it was associated with.

Reply:
Yeah, I was wondering about that too. The conclusion I came to was that the transcript is intimately tied to the bundleID being executed: it doesn't make sense to retain a transcript without also retaining the bundle. It's not like you can replay the concatenated transcripts on top of only the original source bundle and get to the current state.

To rebuild the vat from nothing requires a data structure that looks like "load bundle1, execute these N1 transcript entries, then shutdown and load bundle2, then execute these other N2 entries, etc". To do that properly we'd need to put the bundleID in the `startVat` delivery, and record the `stopVat` deliveries too. And you'd need to retain the source bundles for all the bundleIDs in all the `startVat`s in the extended transcript.

That's not a bad design, but I haven't stopped to think about how long it would take to make the necessary changes to achieve it. I'll spend a bit evaluating that before I land this PR. At the very least I'll sketch out what the complete design would be so I can compare.

Performance-wise, I was hoping to delete as much data as possible, and being able to delete the transcript and the old bundles is appealing. A new validator who's trying to catch up doesn't need to replay the transcripts from bundle-1: it's sufficient for them to replay from just the most recent `startVat`. That helps their catchup performance a lot, especially if we manage periodic "null upgrades" whose main job is to discard the transcripts. But of course we don't have to delete the historical transcripts to achieve this, as long as new validators can merkle-root-validate the most recent transcript segment without needing to fetch all of the earlier ones too (i.e. we need to be judicious with our merkle tree layout).

Reply:
Deleting the old transcript sounds fine to me.  I was just thinking that having continuity of the crank numbers in the historical timeline would be helpful when debugging timelines that cross an update boundary.

Reply:
It looks like it's non-trivial to make this change now. We could change `startVat` to include the bundleID, but the manager would have to snoop that and rewrite the delivery to include the whole bundle (or, better, send a `setBundle` command with the bundle contents before sending the `deliver(startVat)` command). But given the way the supervisor is currently written, that's too late, because the supervisor needs the full bundle to define `buildVatNamespace` before passing it into `makeLiveSlots`. We'd have to change `makeLiveSlots` to *return* a `setBuildVatNamespace`, sort of like how it used to return a `setBuildRootObject`.

I'll open a new ticket to see if we can come back to this later. There's an open policy question as to whether we should retain the historical transcripts from the very beginning of time, or just enough to enable #1691 -style replay (which, to be clear, cannot benefit from replay beyond the most recent version, which is a significant downside of the null-upgrade step).

Reply:
I created #4940 to track this idea.